# Data anonymization

The goal of this ...

## Imports

In [24]:
import numpy as np
import pandas as pd

In [2]:
SEED = 42

## Data

### Loading data

In [139]:
movies_reviews_df = pd.read_parquet("../../data/movies_reviews.parquet")
movies_reviews_df.shape

(171094, 8)

In [140]:
series_reviews_df = pd.read_parquet("../../data/series_reviews.parquet")
series_reviews_df.shape

(35643, 8)

In [141]:
movies_info_df = pd.read_parquet("../../data/movies_info.parquet")
movies_info_df.shape

(984, 43)

In [142]:
series_info_df = pd.read_parquet("../../data/series_info.parquet")
series_info_df.shape

(978, 43)

## Looking at data

### Info

In [143]:
movies_info_df.head()

id                    russian_title              original_title  \
0  535341                       1+1 (2011)                Intouchables   
1   11130  10 причин моей ненависти (1999)  10 Things I Hate About You   
2  606646            12 лет рабства (2013)            12 Years a Slave   
3     502                12 обезьян (1995)              Twelve Monkeys   
4     346    12 разгневанных мужчин (1956)                12 Angry Men   

                                              actors  \
0  [Франсуа Клюзе, Омар Си, Анн Ле Ни, Одри Флеро...   
1  [Хит Леджер, Джулия Стайлз, Джозеф Гордон-Леви...   
2  [Чиветель Эджиофор, Майкл Фассбендер, Лупита Н...   
3  [Брюс Уиллис, Мэделин Стоу, Брэд Питт, Кристоф...   
4  [Генри Фонда, Мартин Болсам, Ли Дж. Кобб, Джоз...   

                                        voice_actors  year  \
0  [Владимир Зайцев, Илья Исаев, Елена Соловьева,...  2011   
1                                                 []  1999   
2  [Денис Беспалый, Илья Исаев, Вероника Саркисов...  2013   
3  [Алексей Иващенко, Марианна Шульц, Всеволод Ку...  1995   
4  [Борис Кордунов, Олег Голубицкий, Михаил Погор...  1956   

               country                          genre  \
0              Франция      драма, комедия, биография   
1                  США      драма, мелодрама, комедия   
2  США, Великобритания      драма, биография, история   
3                  США  фантастика, триллер, детектив   
4                  США      драма, детектив, криминал   

                                              slogan  \
0  «Sometimes you have to reach into someone else...   
1                                «I pine, I perish!»   
2  «The extraordinary true story of Solomon Northup»   
3                «Когда будущее становится историей»   
4  «they have twelve scraps of paper... twelve ch...   

                      director  ... rating_kinopoisk rating_count_kinopoisk  \
0  Оливье Накаш, Эрик Толедано  ...              8.8       1 392 175 оценок   
1                 Джил Джангер  ...              7.8         118 427 оценок   
2                 Стив МакКуин  ...              7.7         221 399 оценок   
3                Терри Гиллиам  ...              7.8         139 294 оценки   
4                  Сидни Люмет  ...              8.5         114 157 оценок   

  rating_imdb rating_count_imdb  \
0  IMDb: 8.50    825 895 оценок   
1  IMDb: 7.30    328 973 оценки   
2  IMDb: 8.10    684 062 оценки   
3  IMDb: 8.00    607 807 оценок   
4  IMDb: 9.00    762 170 оценок   

                                            synopsis world_critics_percentage  \
0  Пострадав в результате несчастного случая, бог...                      75%   
1  Строгий отец поставил условие: Бьянка пойдет н...                      70%   
2  Соломон Нортап был женатым и образованным мужч...                      95%   
3  2035 год... Чудовищный, неизлечимый вирус унич...                      88%   
4  Юношу обвиняют в убийстве собственного отца, е...                     100%   

  world_critics_star_value world_critics_number_of_reviews  \
0                      6.8                      130 оценок   
1                      6.2                       79 оценок   
2                      8.9                      376 оценок   
3                      7.6                       74 оценки   
4                      9.1                       58 оценок   

  russian_critics_percentage russian_critics_number_of_reviews  
0                       100%                         12 оценок  
1                                                               
2                        89%                         27 оценок  
3                                                               
4                          –                          1 оценка  

[5 rows x 43 columns]

For info datasets we are OK.

### Reviews

In [144]:
movies_reviews_df.head()

show_id                username                datetime sentiment  \
0   277328        ilbarskov - 5279   16 февраля 2022 11:21      good   
1   277328  Егор Стрелецкий - 6855    09 ноября 2021 11:47      good   
2   277328              LapinPetrr   21 октября 2021 12:05      good   
3   277328               vEnomanka  22 сентября 2021 23:58       bad   
4   277328               Cooper_HD      08 июля 2021 01:06      good   

                    subtitle  \
0         Достойное зрелище!   
1         «Ломать, колотить»   
2              Лок`Тар огар!   
3  Маляры из студии Близзард   
4       Непонятый новый мир!   

                                         review_body usefulness_ratio  \
0  Варкрафт снял отличный режиссёр Данкан Джонс, ...            0 / 1   
1  По странному стечению обстоятельств Варкрафт н...            2 / 5   
2  Когда-то Blizzard была прекрасной компанией, к...            2 / 3   
3  Среди отрицательных рецензий много критики сце...           6 / 16   
4  Долго я обходил данную площадку в составлении ...            4 / 6   

                                         direct_link  
0  https://www.kinopoisk.ru/user/39222536/comment...  
1  https://www.kinopoisk.ru/user/90478727/comment...  
2  https://www.kinopoisk.ru/user/66435781/comment...  
3  https://www.kinopoisk.ru/user/3351738/comment/...  
4  https://www.kinopoisk.ru/user/4348938/comment/...

For reviews dataset we have to map usernames to user IDs and drop direct link column.  

### Working with usernames

#### Reading

Let's read all usernames under which users left their reviews for both movies and series

In [145]:
movies_reviews_users = movies_reviews_df["username"].values
series_reviews_users = series_reviews_df["username"].values
len(movies_reviews_users), len(series_reviews_users)

(171094, 35643)

#### Merging

Let's merge them, drop duplicates and sort them

In [146]:
users = sorted(list(set(np.concatenate((movies_reviews_users, series_reviews_users)))))
len(users)

69166

#### Checking

In [147]:
users[::1000]

['',
 '\narthurkaf\n',
 '-Barsuk-',
 'Adomaltin',
 'Alice Cosmo',
 'Ann672',
 'Astrid-Flora',
 'Bladezer Lockers Man',
 'Child of Lonely Star',
 'Dania_Lux',
 'DikCinema',
 'Eh_sasha_sasha',
 'FKNSS',
 'Galateya7',
 'Hank Betelgeuse',
 'Imni',
 'Jimenes',
 'KateMoseeva',
 'Krimbley',
 'Leo Naphta',
 'LsySugar',
 'Marina Lemann',
 'Mikhail Khorpyakov',
 'MsOlya',
 'NewTom2015',
 'Omoroge',
 'Pollyyuhan',
 'Reina_del_crepusculo',
 'SUPERNATURALFOREVER',
 'Shakesbeer',
 'Sophie Pashkina',
 'TOX_SIN',
 'TonyAmigos',
 'Vedimir',
 'WildesTier',
 'Zuzuka_zEro',
 'alex_cheredov',
 'antonpan',
 'benjoker',
 'cinema-life-man',
 'densmr',
 'elen_pogodina',
 'fona_t',
 'habhaus',
 'irina2446urakova',
 'katrinaklimina',
 'kritik16',
 'lisnyak.polina',
 'markkil2012',
 'molora',
 'nick_v',
 'p4ggiswon@gmail.com',
 'quoze',
 'samoylovmichael',
 'skvair',
 'sweet_bait',
 'twitter:Anika_tv',
 'vk-20027013:Максим Сопильняк',
 'yammyspot',
 'Александр Толомоненко',
 'Арсений93',
 'Волчица Воды',
 'Дмитри

Empty string is something we want to avoid - let's look at the data

In [148]:
movies_reviews_df[movies_reviews_df["username"] == ""]

show_id username                datetime sentiment  \
766      535341               13 июля 2012 18:23      good   
1060      11130             15 ноября 2009 16:04      good   
21988    252626             14 ноября 2009 14:17      good   
40341    232785           16 сентября 2012 20:39      good   
47502      3561             26 ноября 2009 11:07      good   
51007    280932               28 июля 2009 03:08   neutral   
61069    276376            13 октября 2008 01:31   neutral   
76032       351            24 августа 2008 18:33      good   
80977       389             10 ноября 2009 15:17      good   
81906       739            18 октября 2008 01:42       bad   
90444    273302             27 ноября 2007 19:47      good   
95307    263531               12 июля 2012 21:55      good   
99005    447301               31 июля 2010 14:25      good   
119575   649917             25 ноября 2016 14:51      good   
120174      326                16 мая 2008 21:35      good   
123733   423989               29 июля 2011 18:16       bad   
137658      329                16 мая 2008 14:01      good   
141049   401177             16 ноября 2009 19:28   neutral   
145499   437410            10 августа 2012 22:39      good   
156990   387556             29 ноября 2009 10:23      good   
157095   387556             21 ноября 2009 11:07      good   
167662   420923             07 января 2010 17:32      good   

                                                 subtitle  \
766            «Вот сколько ты стоишь в мире  искусства».   
1060    Твою причёску ненавижу, и то, как говоришь со ...   
21988                                 Большое приключение   
40341                                       Возможно все.   
47502   Она вернулась в его жизнь, как пламя, которое ...   
51007                                                       
61069        Невероятные приключения американцев в Париже   
76032                                                       
80977                  Наемный убийца и маленькая девочка   
81906                                                       
90444                                                       
95307                  «Мы не команда, мы часовая бомба».   
99005                  Сон во сне - тотальное погружение.   
119575                     Гибсон вернулся! И Слава Богу!   
120174                                                      
123733                                                      
137658                                                      
141049  Когда сбываются самые заветные мечты, следует ...   
145499  Все говорили, что  это конец, а это только нач...   
156990                       Хати - навсегда мой герой...   
157095                    Нет повести печальнее на свете…   
167662                        Старый герой в новом свете.   

                                              review_body usefulness_ratio  \
766     Так сказать из Парижа с любовью, французский к...           13 / 5   
1060    Когда впервые увидела этот фильм, сначала поду...           27 / 4   
21988   Фильм-шедевр. Спасибо большое Шону Пенну за эт...           15 / 6   
40341   Этот фильм один из тех, в которых до последней...            8 / 2   
47502   Фильм `Дневник памяти` снят по сюжету одноимен...           24 / 4   
51007   Первые ощущения после просмотра: жизненно, при...           10 / 2   
61069   В последнее время нечасто смотрю боевики в чис...          17 / 14   
76032   Трогательное, смешное, подчас шокирующее путеш...          18 / 14   
80977   Фильм Люка Бессона вышел в далеком 1994 году. ...           19 / 8   
81906   Как ни печально, совершенно мерзкий, бездушный...          53 / 82   
90444   За более чем 25 лет с момента публикации «Тума...          32 / 14   
95307   Мы ждали, ждали и, наконец, дождались! В голов...          23 / 10   
99005   Просмотр фильма, как процесс, очень приятен, а...           33 / 7   
119575  После долгого молчания, длинною в 10 лет, Гибс...           11 / 5   
120174 

We can see that, actually, we have column `direct_link` with link to the comment and this link contains the `user_id`. Different users happen to have empty usernames.  
It will be easier to use actual `user_id`'s than generate them ourselves - we just need to find out how to extract `user_id`'s from links.

#### Extracting `user_id`'s

Some reviews doesn't have direct link:

In [149]:
movies_reviews_df["direct_link"].value_counts()

                                                          6734
https://www.kinopoisk.ru/user/1939008/comment/1449048/       2
https://www.kinopoisk.ru/user/1940813/comment/1449009/       2
https://www.kinopoisk.ru/user/975381/comment/1442974/        2
https://www.kinopoisk.ru/user/278280/comment/1440106/        2
                                                          ... 
https://www.kinopoisk.ru/user/2009568/comment/2612773/       1
https://www.kinopoisk.ru/user/5409448/comment/2599615/       1
https://www.kinopoisk.ru/user/6828711/comment/2569361/       1
https://www.kinopoisk.ru/user/668720/comment/2560131/        1
https://www.kinopoisk.ru/user/1414072/comment/2128039/       1
Name: direct_link, Length: 163988, dtype: int64

So, in this case, it will be easier to replace empty usernames with actual `user_id`'s and then encode usernames to new user IDs.  
This way we will restore information for empty usernames (for building recommendation system).

In [150]:
movies_reviews_df.loc[movies_reviews_df["username"] == "", "username"] = [
    link.split("/")[4]
    for link in movies_reviews_df[movies_reviews_df["username"] == ""][
        "direct_link"
    ].values
]

Let's do the same for series reviews

In [151]:
series_reviews_df[series_reviews_df["username"] == ""]

show_id username               datetime sentiment  \
7467    161268            13 ноября 2009 13:24      good   
9113    453191            09 ноября 2009 13:11      good   
25905   178707           25 октября 2009 22:01      good   

                                        subtitle  \
7467   Мне не до пломбира, пока на попе дыры (с)   
9113                Вампиры, вампиры, вампиры...   
25905                                              

                                             review_body usefulness_ratio  \
7467   Впервые увидела этот мульт лет в 12. Если чест...          32 / 23   
9113   Сейчас пошла просто дикая мода на вампирство.....          48 / 43   
25905  Сериал…сверхъестественное…я по началу думала п...           35 / 3   

                                             direct_link  
7467   https://www.kinopoisk.ru/user/496916/comment/6...  
9113   https://www.kinopoisk.ru/user/496916/comment/6...  
25905  https://www.kinopoisk.ru/user/428435/comment/6...

In [152]:
series_reviews_df.loc[series_reviews_df["username"] == "", "username"] = [
    link.split("/")[4]
    for link in series_reviews_df[series_reviews_df["username"] == ""][
        "direct_link"
    ].values
]

#### Creating map

In [153]:
users = sorted(list(set(np.concatenate((movies_reviews_users, series_reviews_users)))))
len(users)

69172

Before we had 69166 unique users. Now we have 6 more - not much, but better.

In [154]:
user_map = {user: id_ for user, id_ in zip(users, range(len(users)))}

#### Updating dataset

In [155]:
movies_reviews_df["username"] = movies_reviews_df["username"].map(user_map)
series_reviews_df["username"] = series_reviews_df["username"].map(user_map)

Let's also rename column `username` to `user_id`

In [156]:
movies_reviews_df.rename({"username": "user_id"}, axis=1, inplace=True)
series_reviews_df.rename({"username": "user_id"}, axis=1, inplace=True)

And drop `direct_link` and `review_id` columns for anonymized version of the dataset

In [157]:
movies_reviews_df.drop("direct_link", axis=1, inplace=True, errors="ignore")
series_reviews_df.drop("direct_link", axis=1, inplace=True, errors="ignore")

And save the datasets

In [158]:
movies_reviews_df.to_parquet("../../data/anonymized_movies_reviews.parquet")
series_reviews_df.to_parquet("../../data/anonymized_series_reviews.parquet")